riscv/runtime/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+#![no_std]
+#![feature(start)]
+
+use core::arch::asm;
+use core::panic::PanicInfo;
+
+#[panic_handler]
+fn panic(_panic: &PanicInfo<'_>) -> ! {
+    // activate after #176 has been printed
+    // print(format!("panic at line {} in file {}\n", panic.location().line(), panic.location().file()));
+    unsafe {
+        asm!("unimp");
+    }
+    loop {}
+}
+
+#[inline]
+pub fn get_prover_input(index: u32) -> u32 {
+    let mut value: u32;
+    unsafe {
+        asm!("ecall", lateout("a0") value, in("a0") index);
+    }
+    value
+}
+
+extern "Rust" {
+    fn main();
+}
+#[no_mangle]
+#[start]
+pub unsafe extern "C" fn __runtime_start() -> ! {
+    unsafe {
+        main();
+    }
+    loop {}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __runtime_memcpy(dest: *mut u8, src: *const u8, n: usize) -> *mut u8 {
+    if (src as u32) % 4 == 0 && (dest as u32) % 4 == 0 {
+        memcpy_aligned(dest, src, n);
+    } else {
+        for i in 0..n {
+            *dest.offset(i as isize) = *src.offset(i as isize);
+        }
+    }
+    dest
+}
+
+pub unsafe fn memcpy_aligned(dest: *mut u8, src: *const u8, n: usize) {
+    let mut i: isize = 0;
+    while i + 3 < n as isize {
+        *((dest.offset(i)) as *mut u32) = *((src.offset(i)) as *mut u32);
+        i += 4;
+    }
+    if i < n as isize {
+        let value = *((src.offset(i)) as *mut u32);
+        let dest_value = (dest.offset(i)) as *mut u32;
+        let mask = (1 << (((n as isize - i) * 8) as u32)) - 1;
+        *dest_value = (*dest_value & !mask) | (value & mask);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __runtime_memset(dest: *mut u8, val: u8, n: usize) -> *mut u8 {
+    let mut i: isize = 0;
+    let pattern = (val as u32) << 24 | (val as u32) << 16 | (val as u32) << 8 | (val as u32);
+
+    // Prologue: set bytes in the first non-aligned word
+    let first_word_offset = (dest as u32) % 4;
+    if first_word_offset != 0 {
+        let aligned_dest = dest.offset(-(first_word_offset as isize)) as *mut u32;
+        let mask = (1 << ((first_word_offset * 8) as u32)) - 1;
+        *aligned_dest = (*aligned_dest & mask) | (pattern & !mask);
+        i += 4 - first_word_offset as isize;
+    }
+
+    // Set all full words
+    while i + 3 < n as isize {
+        *((dest.offset(i)) as *mut u32) = pattern;
+        i += 4;
+    }
+
+    // last part: set the final word
+    if i < n as isize {
+        let dest_value = (dest.offset(i)) as *mut u32;
+        let mask = (1 << (((n as isize - i) * 8) as u32)) - 1;
+        *dest_value = (*dest_value & !mask) | (pattern & mask);
+    }
+
+    dest
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __runtime_memcmp(dest: *mut u8, src: *const u8, n: usize) -> bool {
```

Comment:
Shouldn't this return an `int` like `stcmp`?

Reply:
Maybe we can first merge a byte-by-byte comparison function?

Reply:
Yeah, `memcmp::Memcmp` returns `bool` but indeed the compiler builtin retuns an `i32`. I'll change that (and will comment out what isn't byte-by-byte comparison if you prefer).